dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java

```diff
@@ -1352,25 +1352,43 @@ public void updateReferenceType(String repositoryId, Version version) {
 
     @Override
     protected String getCommitID(String repositoryId, Version version) {
+        try {
+            return getCommitID(repositoryId, version.getReference(), true);
+        } catch (CustomWebApplicationException ex) {
+            return null;
+        }
+    }
+
+    /**
+     * Get the head commit SHA of a specified reference in a particular repository.
+     * @param repositoryId name of the repository (ex 'svonworl/test-notebooks')
+     * @param commitReference reference
+     * @param stripped if true, commitReference does not include the 'refs/{types}/' prefix
+     */
+    public String getCommitID(String repositoryId, String commitReference, boolean stripped) {
         GHRepository repo;
         try {
             repo = github.getRepository(repositoryId);
             GHRef[] refs = getBranchesAndTags(repo);
```

Comment:
This was like this, but I worry it could be a lot more API requests now; I think this will get invoked every time there is a push, whereas it might not have before? I could be wrong on that though. If we're already doing this on every push, then ignore, this comment.

With API caching, it may not matter that much against the limit, but we still risk, with one of the repos that has ~700 branches, let's say 350x3 GitHub API invocations within a minute, if we get 3 events in a short time frame.

It looks like the branch fetching could be optimized fairly easily -- there's a `GHRepository.getBranch(<branchName>)`, but the tags might still have to be  iterated.

I'm just looking for any potential flaws in the PR, and that's the only one I've noticed. I could be off, and even if it is a lot, it could be done in a followup ticket.

Reply:
Looking at the [current] kohsuke library, `getBranchesAndTags` calls `repo.getRefs`, which deep down calls `GHRef.readMatching`, which hits the API endpoints `https://api.github.com/repos/<org>/<repo>/git/refs/tags` and `https://api.github.com/repos/<org>/<repo>/git/refs/heads` to grab the tags and branches, all of which seem to be returned in a single response, try:
```
curl https://api.github.com/repos/broadinstitute/warp/git/refs/tags
```
If that's true, and I counted right, `getCommitID` incurs 4 GitHub calls: one to get the repo info, two for the branches and tags, and one to retrieve the SHA from the matching ref.  And `getCommitID` is called once to check whether or not the push should be processed. 

So, maybe we're ok, but if we're not convinced, we can instrument the code and experiment with the warp repo to be sure.

Reply:
> If that's true, and I counted right, `getCommitID` incurs 4 GitHub calls: one to get the repo info, two for the branches and tags, and one to retrieve the SHA from the matching ref. And `getCommitID` is called once to check whether or not the push should be processed.

Yikes, I blew that calculation; I'd even looked at the code before writing my comment, but somehow still ended up in my head with 1 API call per branch.

You're right about curl returning everything; I'm confused by that. Per their [doc](https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28), that API should be paginated and should return only 30 results by default. If it is really paginating, then it should be about 25 API calls, which seems reasonable

Reply:
> > If that's true, and I counted right, `getCommitID` incurs 4 GitHub calls: one to get the repo info, two for the branches and tags, and one to retrieve the SHA from the matching ref. And `getCommitID` is called once to check whether or not the push should be processed.
> 
> Yikes, I blew that calculation; I'd even looked at the code before writing my comment, but somehow still ended up in my head with 1 API call per branch.
> 
> You're right about curl returning everything; I'm confused by that. Per their [doc](https://docs.github.com/en/rest/branches/branches?apiVersion=2022-11-28), that API should be paginated and should return only 30 results by default. If it is really paginating, then it should be about 25 API calls, which seems reasonable

Turns out they're different endpoints, and indeed, a quick `curl` confirms that the latter is paged:
```
https://api.github.com/repos/broadinstitute/warp/git/refs/tags
https://api.github.com/repos/broadinstitute/warp/branches
```
I think kohsuke uses the former unpaged one, though.

Reply:
If rate limiting is a concern, `getCommitId` could be optimized because we know the git reference name and the git reference type (tag or branch) so we could call the endpoint to get the specific git reference directly instead of iterating through them all
- https://docs.github.com/en/rest/git/refs?apiVersion=2022-11-28#get-a-reference

Reply:
To soothe any concerns, I created a new `getCommitID` function which uses the above endpoint internally, and reverted the old `getCommitID` implementation, so the existing code will work the same as before.  Not sure it was worth the time, I'm pretty sure it only saves one (1) API call per push, although it is easier on the GitHub servers (no need to enumerate the tags and branches), and reduces the overally processing latency, at least a little. 

The old `getCommitID` implementation could be recoded to use the new function, but I was wary of doing do, because the code feels like it'd be easy to break, and there's no pressing need...